source/base/function_signed_distance.cc

```diff
@@ -353,6 +353,166 @@ namespace Functions
+      // distance to the bottom edge of the notch (2D: minimum distance to the
+      // two corner points; 3D: minimum distance to codim-1 sphere)
+      double distance_bottom_notch_edge =
+        std::min(p.distance(notch.vertex(0)), p.distance(notch.vertex(1)));
+
+      // 3D: Create codim-1 spheres for detection of edges
+      if (dim == 3)
+        {
+          const unsigned int codim = std::max(1, dim - 1);
+          const unsigned int y_idx = std::min(1, dim);
+
+          // distance to bounding circles lying y-z plane
+          double r = std::sqrt(std::pow(radius, 2) -
+                               std::pow(0.5 * notch.side_length(0), 2));
+          Functions::SignedDistance::Sphere<codim> circle(
+            Point<codim>(center_sphere[y_idx], center_sphere[dim - 1]), r);
+
+          distance_bottom_notch_edge = std::sqrt(
+            std::pow(circle.value(Point<codim>(p[y_idx], p[dim - 1])), 2) +
+            std::pow(signed_distances_notch[0], 2));
+
+          // distance to bounding circles lying in x-y plane
+          r = std::sqrt(std::pow(radius, 2) -
+                        std::pow(notch.side_length(dim - 1) - radius, 2));
+          Functions::SignedDistance::Sphere<codim> circle_top(
+            Point<codim>(center_sphere[0], center_sphere[y_idx]), r);
+
+          distance_front_notch_edge = std::sqrt(
+            std::pow(circle_top.value(Point<codim>(p[0], p[y_idx])), 2) +
+            std::pow(signed_distances_notch[idx_top], 2));
+        }
+
+      // point is within bounds of the notch faces
+      if (*std::min_element(signed_distances_notch.begin(),
+                            signed_distances_notch.end()) > 0)
+        {
+          const double r =
+            std::sqrt(std::pow(radius, 2) -
+                      std::pow(notch.side_length(dim - 1) - radius, 2));
+
+          // inside notch
+          if ((dim == 2 && notch.point_inside(p)) ||
+              (dim == 3 && p.distance(center_sphere) <= radius))
+            {
+              return std::min(
+                distance_front_notch_edge,
+                std::min(distance_bottom_notch_edge,
+                         *std::min_element(signed_distances_notch.begin(),
+                                           signed_distances_notch.end())));
+            }
+          // case: below top face of notch
+          else if (dim == 2 ||
+                   (dim == 3 && (std::abs(p[1] - center_sphere[1])) < r))
+            {
+              return std::min(distance_bottom_notch_edge,
+                              signed_distances_notch[idx_top]);
+            }
+          // case: corner, if the point is outside sphere but within notch
+          // bounds
+          else
+            return std::min(distance_bottom_notch_edge,
+                            distance_front_notch_edge);
+        }
+      // case: inside disk
+      else if (sphere.value(p) <= 0)
+        {
+          return std::max(sphere.value(p), -notch.signed_distance(p));
```

Comment:
I am super-confused about the implementation of the `value`-function. What you do on this line is a setminus operation with level set functions:
![image](https://user-images.githubusercontent.com/9609727/230485203-28ccd72f-4d2d-417d-a71d-16955aad9f74.png)

(table from http://doi.wiley.com/10.1002/nme.4823).
But then it looks to me like the whole `value`-function could be just this line. Why do you need the rest?

Reply:
Thanks a lot, I will have a look at it asap.

Reply:
In our case, the setminus operation results in the correct signed distance function for points lying inside the disk but not necessarily for points outside the disk. This is due to the following reasons: `notch` actually represents the signed distance function to the bounding box of the notch, i.e., the signed distance function to a closed hyperrectangle. Thus, for points lying outside the disk and in the vicinity of the corner points of the notch (2D: zones 6,7,10 and 11 in the drawing below taken from _Mousavi, R. (2014). Level set method for simulating the dynamics of the fluid-fluid interfaces: Application of a Discontinuous Galerkin Method_)
<img src="https://user-images.githubusercontent.com/70260016/231160532-8b53be3f-672a-4b49-a11d-af52306aba0a.png" width=50% height=50%>

we need special treatment to get the correct signed distance function. To this end, I introduced the cases in the remaining lines https://github.com/dealii/dealii/blob/d332f08c96e4de5804f2216494eade2ee13af0ce/source/base/function_signed_distance.cc#L459-L489 and 
https://github.com/dealii/dealii/blob/d332f08c96e4de5804f2216494eade2ee13af0ce/source/base/function_signed_distance.cc#L495-L514.

I have to admit that the name `notch` may be a bit misleading here, and maybe that's what led to your confusion. 
Since I went through the code lines again and found that they are not very easy to follow unless you are directly involved in the implementation, I will try to come up with a more concise and readable implementation as soon as possible.

Reply:
I'm afraid I still don't understand why you need the cases. If you just make the notch, so that it extends outside the disk:
![image](https://user-images.githubusercontent.com/9609727/232098931-8e2f0f46-c0cc-4e3a-b25a-c461b5367d9f.png)
I don't think there is any problem with setminus. If you compare the implementation you have now with only using setminus it looks like this:

![image](https://user-images.githubusercontent.com/9609727/232099150-4e4b8831-78f8-49a7-bf40-bb81cc2d1132.png)

And I really believe the setminus will give the correct sign-distance function, because its gradient will have norm equal to 1, if the two functions you construct it from have gradients with norm equal to 1:
![image](https://user-images.githubusercontent.com/9609727/232099964-3f01f4db-1ef5-428f-bdac-91ff412f530e.png)

Reply:
Convinced. I really could not believe it was that easy, but I am glad it is. Thank you @Simon for your efforts. I added your name to the changelog file, I hope that is okay with you as well.